crates/oxc_transformer/src/decorator/legacy/metadata.rs

```diff
@@ -201,6 +224,71 @@ impl<'a> Traverse<'a, TransformState<'a>> for LegacyDecoratorMetadata<'a, '_> {
 }
 
 impl<'a> LegacyDecoratorMetadata<'a, '_> {
+    /// Collects enum type information for decorator metadata generation.
+    fn collect_enum_type(&mut self, decl: &TSEnumDeclaration<'a>, ctx: &TraverseCtx<'a>) {
+        let symbol_id = decl.id.symbol_id();
+
+        // Optimization:
+        // If the enum doesn't have any type references, it is said that no decorator will
+        // refer to this enum, so there is no need to infer its type.
+        let has_type_reference =
+            ctx.scoping().get_resolved_references(symbol_id).any(Reference::is_type);
+        if has_type_reference {
+            let enum_type = Self::infer_enum_type(&decl.body.members);
+            self.add_enum_type(symbol_id, enum_type);
+        }
+    }
+
+    fn add_enum_type(&mut self, symbol_id: SymbolId, enum_type: EnumType) {
+        self.enum_types.insert(symbol_id, enum_type);
+    }
+
+    /// Infer the type of an enum based on its members
+    fn infer_enum_type(members: &[TSEnumMember<'a>]) -> EnumType {
+        let mut has_string = false;
+        let mut has_number = false;
+
+        for member in members {
+            if let Some(init) = &member.initializer {
+                match init {
+                    Expression::StringLiteral(_) if !has_number => {
+                        has_string = true;
+                    }
+                    Expression::TemplateLiteral(literal)
+                        if !has_number
+                            && literal
+                                .expressions
+                                .iter()
+                                .all(|expr| matches!(expr, Expression::StringLiteral(_))) =>
+                    {
+                        has_string = true;
+                    }
+                    Expression::NumericLiteral(_) | Expression::BigIntLiteral(_) if !has_string => {
+                        has_number = true;
+                    }
+                    Expression::UnaryExpression(expr)
+                        if expr.argument.is_number_literal() && !has_string =>
+                    {
+                        has_number = true;
+                    }
```

Comment:
Not all `UnaryExpression`s evaluate to a number. It depends on the operator:

* `!123` -> boolean
* `typeof 123` -> string
* `void 123` -> undefined
* `delete 123` -> boolean

The first one is not entirely stupid - `!0` and `!1` are sometimes used as shortcuts for `true` / `false`.